Rainjacked/rainscript
# Parser Utility - Linked List and Environment

# Summary

Propose for the two utility classes (LinkedList and Environment) to be used for variable scoping during compilation of Rainscript.

## Persistent Linked List

The linked lists here are persistent, hence is most similar to racket language's `cons`. This will be used to persistently log warnings/errors during recursive descent.

## Persistent Environment

The Environment will be similar to the linked list, however keeps both a key and a value instead of a single data head. Though the two are similar, they will have no superclass/subclass relationship since they handle their methods differently. The Environment will mainly be used for variable identifier lookups and block-scoping in Rainscript.

The `push` method cheaply creates a new Environment every time. This style is used so that it is easier to `reset` block scopes during recursive-descent in, for example, `#if` or `#choice` blocks.

> Note: Environment is NOT for variable content but for data types, since this is just the compiler.

# How to Test

1. Have a fresh clone of this PR by running:
```
git clone https://github.com/Hikari9/rainscript -b parser-00-util rainscript-parser-00-util
cd rainscript-parser-00-util
```
  Alternatively, you may also fetch using a remote to `Hikari9/parser-00-util`.

2. For first use, make sure npm is installed then run:
```
npm install
```

3. To perform tests on your machine, run:
```
npm test
```

4. (Optional) Add your own tests to `/tests/parser/util/LinkedList.js` or `/tests/parser/util/Environment.js`. Push the new tests into this PR so that it can be reviewed (if you are a maintainer). Otherwise simply leave a comment.
